Implements #935 - Insert IPv4 and IPv6 call in /networks/id/connect

mdaue opened the pull request:
Requires API 1.22

Unit tests passed, had some trouble running integration tests:
E   APIError: 400 Client Error: Bad Request ("client is newer than server (client API version: 1.22, server API version: 1.21)")

Docker version:
Client:
 Version:      1.11.0-dev
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   415dd86
 Built:        Sun Feb  7 21:47:30 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.0-dev
 API version:  1.23
 Go version:   go1.5.3
 Git commit:   415dd86
 Built:        Sun Feb  7 21:47:30 2016
 OS/Arch:      linux/amd64

dnephin commented on docker/api/network.py:
```diff
@@ -46,9 +46,10 @@ def inspect_network(self, net_id):
 
     @check_resource
-    @minimum_version('1.21')
+    @minimum_version('1.22')
```
This isn't correct, the function needs to stay at `1.21`. You can add an `if` branch within the function to require a higher version for specific params (there are examples of this already).

dnephin commented on tests/integration/network_test.py:
```diff
@@ -100,7 +100,7 @@ def test_remove_network(self):
         self.assertEqual(len(self.client.networks()), initial_size)
 
-    @requires_api_version('1.21')
+    @requires_api_version('1.22')
```
same here

dnephin commented on tests/integration/network_test.py:
```diff
@@ -100,7 +100,7 @@ def test_remove_network(self):
-    @requires_api_version('1.21')
+    @requires_api_version('1.22')
     def test_connect_and_disconnect_container(self):
         net_name, net_id = self.create_network()
```
same here

dnephin commented:
Thanks for the PR. I left aa couple comments.

Once it's ready to go, we'll also ask you to squash your commits down to one or two.

mdaue commented:
Great! I squashed to one commit, leaving interesting comments in the notes. I set the minimum required API back to 1.21 where identified and compared with master before committing. I removed my addition to the unit test as it failed, possibly because it wasn't run with API 1.22. Alternatively, I can add that test again and perform the same version check.

Thanks!

dnephin commented on docker/api/network.py:
```diff
@@ -56,6 +58,16 @@ def connect_container_to_network(self, container, net_id,
         }
+
+        # IPv4 or IPv6 or neither:
+        if version_gte(self._version, '1.22'):
```
I think this needs to be the other way around. 

```python
if version_lt(...):
    raise errors.InvalidVersion(...)

```

mdaue commented:
Just in case it was missed, I did update the code base on the suggestions and squashed the commit.

shin- commented on docker/api/network.py:
```diff
@@ -56,6 +59,19 @@ def connect_container_to_network(self, container, net_id,
+        if ip4 or ip6:
+            if version_lt(self._version, '1.22'):
+                raise InvalidVersion('IP address assignment is not in '
+                                     'supported in API ver < 1.2')
```
Let's spell out "version" entirely 
+ typo `1.22`

shin- commented on docker/api/network.py:
```diff
@@ -48,7 +50,8 @@ def inspect_network(self, net_id):
     @minimum_version('1.21')
     def connect_container_to_network(self, container, net_id,
-                                     aliases=None, links=None):
+                                     ip4=None, ip6=None, aliases=None,
```
Let's call those `ipv4_address` an `ipv6_address`. The additional characters are worth the clarity they provide, I think.

shin- commented:
A few small comments, should be good after that.

mdaue commented:
Thanks - made changes.

shin- commented:
LGTM :+1:

dnephin commented on tests/integration/network_test.py:
```diff
@@ -235,3 +235,33 @@ def test_connect_with_links(self):
+        self.assertEqual(
+            container_data['NetworkSettings']['Networks'][net_name]
+                          ['IPv6Address'],
+            '2001:389::1')
```
Minor: This could probably be two separate tests

dnephin commented:
LGTM

mdaue commented:
I split the integration tests as per @dnephin suggestion. Also found that I had forgotten to change the arg to the function calls in the test (ip4 vs. ipv4_address) after making the change in network.py (fixed).

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "935" git@github.com:mdaue/docker-py.git somewhere
$ cd somewhere
$ git rebase -i HEAD~4
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "935" git@github.com:mdaue/docker-py.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.

GordonTheTurtle commented:
Please sign your commits following these rules:
https://github.com/docker/docker/blob/master/CONTRIBUTING.md#sign-your-work
The easiest way to do this is to amend the last commit:
~~~console
$ git clone -b "935" git@github.com:mdaue/docker-py.git somewhere
$ cd somewhere
$ git rebase -i HEAD~2
editor opens
change each 'pick' to 'edit'
save the file and quit
$ git commit --amend -s --no-edit
$ git rebase --continue # and repeat the amend for each commit
$ git push -f
~~~

Ammending updates the existing PR. You **DO NOT** need to open a new one.
